tests/accumulate-headers-after-304.js

```diff
@@ -0,0 +1,178 @@
+
+        let updateField = new Field("X-Update-Header", Gadgets.UniqueId("update"));
+        updateField.finalize();
+        const updateHeaderLength = updateField.raw().length;
+        
+        {
+            let testCase = new HttpTestCase('forward a cachable response');
+            testCase.client().request.for(resource);
+            if (Config.smp())
+                testCase.client().nextHopAddress = this._workerListeningAddresses[1];
+            testCase.server().serve(resource);
+            let startLine = testCase.server().response.startLine;
+            startLine.finalize(); // 200 by default
+            const headerSeparator = 2; // "\r\n"
+            const finalPrefixSize = headerSize + startLine.raw().length + 2; // will become ofter receiving 304 updating response
+            const initialPrefixSize = finalPrefixSize - updateHeaderLength;
+            assert(initialPrefixSize <= MaxHeaderSize);
+            testCase.server().response.enforceMinimumPrefixSize(initialPrefixSize);
+            testCase.server().response.header.add(hitCheck);
+            await testCase.run();
+        }
+
+        {
+            let testCase = new HttpTestCase('respond with a 304 hit');
+            testCase.client().request.for(resource);
+            if (Config.smp())
+                testCase.client().nextHopAddress = this._workerListeningAddresses[1];
+            testCase.client().request.conditions({ ims: resource.notModifiedSince() });
+            testCase.client().checks.add((client) => {
+                client.expectStatusCode(304);
+            });
+            await testCase.run();
+        }
+
+        let updatingResponse = null;
+        {
+            let testCase = new HttpTestCase(`get a 304 that increases the cached header size on ${updateHeaderLength}`);
+
+            resource.modifyNow();
+            resource.expireAt(FuzzyTime.DistantFuture());
+
+            testCase.client().request.for(resource);
+            testCase.client().request.conditions({ ims: resource.modifiedSince() });
+            testCase.client().request.header.add("Cache-Control", "max-age=0");
+            if (Config.smp())
+                testCase.client().nextHopAddress = this._workerListeningAddresses[2];
+
+            testCase.server().response.header.add(updateField);
+            testCase.server().response.startLine.code(304);
+            testCase.server().serve(resource);
+            testCase.check(() => {
+                const receivedResponse = testCase.client().transaction().response;
+                const code = receivedResponse.startLine.codeInteger();
+                updatingResponse = testCase.server().transaction().response;
+                if (headerSize <= MaxHeaderSize) {
+                    assert.equal(updatingResponse.id(), receivedResponse.id(), "relayed X-Daft-Response-ID");
+                    assert.equal(code, 200);
+                } else {
+                    assert(code>=500);
```

Comment:
Hmm. https://github.com/measurement-factory/squid/pull/71 promises to retry the revalidation request but this code does not seem to test that part (because the Daft server stops listening after answering the revalidation request?). Can we test the _whole_ expected sequence, including a successful miss transaction (that gets successfully cached!). I would not add a hitCheck header to that post-revalidation miss transaction, so that we can be (more) sure that Squid got rid of the response that grew a too-large header.

Reply:
Unfortunately I could not find a way how to make the server serve the second response.

Reply:
I managed to figure out how to implement this: I had to modify server agent's keepListening()  functionality. I posted these changes in SQUID-887-304-headers-accumulation (created from SQUID-479-large-header with the recent master changes and this adjustment).